tensorflow/core/common_runtime/zen_layout_pass.cc

```diff
@@ -0,0 +1,1196 @@
+        zen_nodes.push_back(n);
+      }
+    }
+  }
+
+  std::unordered_map<Node *, std::pair<bool, bool>> reorder_flags =
+      GetReorderFlags(zen_nodes);
+
+  for (Node *n : zen_nodes) {
+    std::string node_name = n->name();
+    std::string op_name = n->type_string();
+    std::pair<bool, bool> n_reorder = reorder_flags[n];
+
+    ZenOpRewriteRecord rewrite_record;
+    for (auto it = zen_rewrite_db_.begin(); it < zen_rewrite_db_.end(); it++) {
+      if (op_name == it->zen_op_name) {
+        rewrite_record = *it;
+        break;
+      }
+    }
+
+    // Rewrite op with a copy containing the new reorder flags.
+    if (ZenOpNodeRewrite(g, n, &rewrite_record, n_reorder) == OkStatus()) {
+      VLOG(1) << "ZenLayoutRewritePass::AddReorderAttrs: Node " << node_name
+              << " " << op_name << " updated reorders to " << n_reorder.first
+              << " " << n_reorder.second;
+      result = true;
+    }
+  }
+
+  return result;
+}
+
+bool ZenLayoutRewritePass::ZenOpUpdate(std::unique_ptr<Graph> *g) {
+  bool result = false;
+  std::vector<Node *> order;
+  GetReversePostOrder(**g, &order);
+  for (Node *n : order) {
+    if (!n->IsOp() || !CanOpRunOnCPUDevice(n)) {
+      continue;
+    }
+
+    const ZenOpRewriteRecord *rewrite_record = nullptr;
+    if ((rewrite_record = CheckNodeForZenOpRewrite(n)) != nullptr) {
+      string node_name = n->name();
+      string op_name = n->type_string();
+      std::pair<bool, bool> n_reorder(true, true);
+      if (ZenOpNodeRewrite(g, n, rewrite_record, n_reorder) == OkStatus()) {
+        VLOG(1) << "ZenLayoutRewritePass::ZenOpUpdate: Node " << op_name
+                << " rewritten with ZenOp " << rewrite_record->zen_op_name;
+        result = true;
+      } else {
+        // Rewriting the node with Zen op failed. Hence the existing node will
+        // be there with graph for inference.
+        VLOG(1) << "ZenLayoutRewritePass::ZenOpUpdate: Failed to rewrite node "
+                << node_name << " with Zen op " << op_name;
+      }
+    }
+  }
+  return result;
+}
+
+// Method to find whether the graph has inference ops only. It returns error
+// status if the graph has training ops.
+Status ZenLayoutRewritePass::AreAllInferenceOps(std::unique_ptr<Graph> *g) {
+  std::vector<Node *> order;
+  GetReversePostOrder(**g, &order);
+  for (Node *n : order) {
+    if (!n->IsOp()) {
+      continue;
+    }
+    for (auto op = tf_training_ops_.cbegin(); op != tf_training_ops_.cend();
+         ++op) {
+      if (n->type_string().find(*op) != string::npos) {
+        return Status(error::Code::UNIMPLEMENTED,
+                      "Training operation found! Currently TF-ZenDNN "
+                      "does not support training. Set environment "
+                      "variable TF_ENABLE_ZENDNN_OPTS to '0' for training.");
+      }
+    }
+  }
+  return OkStatus();
+}
+
+bool ZenLayoutRewritePass::ZenOpRewritePass(std::unique_ptr<Graph> *g) {
+  bool result = false;
+  CHECK_NOTNULL(g);
```

Comment:
```suggestion
  CHECK_NOTNULL(g);  // Crash ok.
```
Same for other places.

Reply:
Fixed. Thank you!